components/prize_pool/commons/prize_pool.lua

```diff
@@ -0,0 +1,199 @@
+		end,
+		row = 'points'
+	},
+	[prizeTypes.QUALIFIES] = {
+		header = 'qualifies',
+		headerParse = function (input)
+			-- TODO: Add manual and automatic retrevial of additional data
+			return {link = input:gsub(' ', '_')}
+		end,
+		row = 'qualified'
+	},
+	[prizeTypes.FREETEXT] = {
+		header = 'freetext',
+		headerParse = function (input)
+			return {title = input}
+		end,
+		row = 'freetext'
+	},
+}
+
+function PrizePool:init(args)
+	self.args = args
+	self.pagename = mw.title.getCurrentTitle().text
+	self.args.type = Json.parseIfString(self.args.type)
+
+	if not self.args.type then
+		return error('Please provide a type!')
+	elseif type(self.args.type) ~= 'table' or not self.args.type.type then
+		return error('Could not parse type!')
+	elseif not Opponent.isType(self.args.type.type) then
+		return error('Not a valid type!')
+	end
+
+	self.type = self.args.type.type
+
+	self.options = self:_readStandardOptions(args)
+	self.prizes = self:_readStandardPrizes(args)
+	if self.options.showUSD then
+		self:addPrize('USD', 1)
+	end
+	self.placements = self:_readPlacements(args)
+end
+
+function PrizePool:create()
+	-- TODO
+	mw.logObject(self)
+	return ''
+end
+
+function PrizePool:_readStandardOptions(args)
+	self.options = {}
+
+	self:setOption('autoUSD', Logic.nilOr(Logic.readBoolOrNil(args.autousd), true))
+	self:setOption('showUSD', Logic.nilOr(Logic.readBoolOrNil(args.showusd), true)) -- TODO: Improve
+	self:setOption('prizeSummary', Logic.nilOr(Logic.readBoolOrNil(args.prizesummary), true))
+	self:setOption('exchangeInfo', Logic.nilOr(Logic.readBoolOrNil(args.exchangeinfo), true))
+	self:setOption('storeSmw', Logic.nilOr(Logic.readBoolOrNil(args.storesmw), true))
+	self:setOption('storeLpdb', Logic.nilOr(Logic.readBoolOrNil(args.storelpdb), true))
+
+	return self.options
+end
+
+function PrizePool:_readStandardPrizes(args)
+	self.prizes = {}
+
+	for prizeEnum in pairs(prizeTypes) do
+		local prizeDatum = prizeData[prizeTypes[prizeEnum]]
+		local fieldName = prizeDatum.header
+		if fieldName then
+			args[fieldName .. '1'] = args[fieldName .. '1'] or args[fieldName]
+			for _, prizeValue, index in Table.iter.pairsByPrefix(args, fieldName) do
+				local data = prizeDatum.headerParse(prizeValue)
+				self:addPrize(prizeEnum, index, data)
+			end
+		end
+	end
+
+	return self.prizes
+end
+
+function PrizePool:_readPlacements(args)
+	self.placements = {}
+
+	local currentPlace = 0
+	for placementIndex = 1, math.huge do
+		local placementInput = Json.parseIfString(args[placementIndex])
+		if not placementInput then
+			break
+		end
+
+		local placement = {}
+		-- TODO: Parse prizes
+		placement.prizes = {}
+
+		placement.opponents = {}
+		for opponentIndex = 1, math.huge do
+			if not placementInput[opponentIndex] then
+				break
+			end
+			local parsedOpponent = Json.parseIfString(placementInput[opponentIndex])
+			table.insert(placement.opponents, parsedOpponent)
+			-- TODO: parse it into our struct
```

Comment:
do i assume correctly that here we will also sort prize, points etc into the opponents?

Reply:
There will be a way to customize it, but there won't be any overridable functions (as they aren't a good design choice imo)

Reply:
basically what i would want to do is execute the following after `opponent.opponentData = Opponent.readOpponentArgs(opponentInput) or Opponent.tbd()`:

```lua
	local partySize = Opponent.partySize(opponent.opponentData.type)

	if partySize == 1 then
		opponent.opponentData.players[1].race = StarcraftRace.read(opponentInput.race)

	elseif partySize then
		opponent.opponentData.isArchon = Logic.readBool(opponentInput.isarchon)
		if opponent.opponentData.isArchon then
			local archonRace = StarcraftRace.read(opponentInput.race)
			for _, player in ipairs(opponent.opponentData.players) do
				player.race = archonRace
			end
		else
			for playerIndex, player in ipairs(opponent.opponentData.players) do
				player.race = StarcraftRace.read(opponentInput['p' .. playerIndex .. 'race'])
			end
		end
	end
```

Reply:
What you want is to run `Module:Opponent/Starcraft``.readOpponentArgs()` instead of `Module:Opponent``.readOpponentArgs()` correct?

Reply:
basically yeah